File: vw_rust/src/render_pass/attachment.rs
use crate::sys::bindings::{VwAttachment, VwString, VwFormat, VwImageLayout};
use std::ffi::CString;

pub struct Attachment {
    id: CString,
    format: VwFormat,
    final_layout: VwImageLayout,
}

pub struct AttachmentBuilder {
    id: CString,
    format: VwFormat,
    final_layout: VwImageLayout,
}

impl AttachmentBuilder {
    pub fn new(id: &str) -> AttachmentBuilder {
        AttachmentBuilder {
            id: CString::new(id).unwrap(),
            format: VwFormat::B8G8R8A8_SRGB,
            final_layout: VwImageLayout::Undefined,
        }
    }

    pub fn with_format(mut self, format: VwFormat) -> AttachmentBuilder {
        self.format = format;
        self
    }

    pub fn with_final_layout(mut self, layout: VwImageLayout) -> AttachmentBuilder {
        self.final_layout = layout;
        self
    }

    pub fn build(self) -> Attachment {
        Attachment {
            id: self.id,
            format: self.format,
            final_layout: self.final_layout,
        }
    }
}

impl Attachment {
    pub fn to_raw(&self) -> VwAttachment {
        VwAttachment {
            id: VwString{string: self.id.as_ptr()},
            format: self.format,
            finalLayout: self.final_layout,
        }
    }
}

File: vw_py/RenderPass/Attachment.py
import bindings_vw_py as bindings

class Attachment:
    def __init__(self, id, format, final_layout):
        self.id = bindings.CString(id)
        self.attachment = bindings.VwAttachment()
        self.attachment.finalLayout = final_layout
        self.attachment.format = format
        self.attachment.id = bindings.VwString(self.id)

class AttachmentBuilder:
    def __init__(self, id):
        self.id = id
        self.format = 0
        self.final_layout = 0

    def with_final_layout(self, layout):
        self.final_layout = layout
        return self
    
    def with_format(self, format):
        self.format = format
        return self
    
    def build(self):
        return Attachment(self.id, self.format, self.final_layout)
